lib/core/test/unit/Cardano/Wallet/StakePool/MetricsSpec.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Cardano.Wallet.StakePool.MetricsSpec (spec) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types
+    ( BlockHeader (..)
+    , EpochLength (..)
+    , Hash (..)
+    , PoolId (..)
+    , SlotId (..)
+    , flatSlot
+    , fromFlatSlot
+    )
+import Cardano.Wallet.StakePool.Metrics
+    ( State (..), applyBlock )
+import Data.Quantity
+    ( Quantity (..) )
+import Numeric.Natural
+    ( Natural )
+import Test.Hspec
+    ( Spec, describe, it, shouldBe )
+import Test.QuickCheck
+    ( Arbitrary (..), InfiniteList (..), property )
+import Test.QuickCheck.Arbitrary.Generic
+    ( genericArbitrary, genericShrink )
+
+import qualified Data.ByteString as BS
+
+spec :: Spec
+spec = do
+    describe "Counting how many blocks each pool produced" $
+        describe "State" $ do
+            it "stores the last applied blockHeader"
+                $ property $ \s b@(header,_) -> do
+                tip (applyBlock b s) `shouldBe` header
```

Comment:
we can also construct property: 
- after a series of `applyBlock` we will have unique values (no-repeating)
- after a series of `applyBlock` we will have all slotId values (no holes)
- after a series of `applyBlock` values for a given PoolId cannot shrink, can only keep growing
...

Reply:
Thanks!! 

I added a variant of "after a series of applyBlock we will have all slotId values (no holes)" that checks that the total activity count increases by `+1` when a block is applied.

I didn't attempt the other ones now. If you do want to, feel free. (Or I could get to it later)

> after a series of applyBlock we will have unique values (no-repeating)

This would require generating a sound sequence of arbitrary blocks… But this might be very good to have for rollbacks to make sure we don't count a slot twice 🤔 

> after a series of applyBlock values for a given PoolId cannot shrink, can only keep growing

When we are going to support rollback, `applyBlock` will take some kind of `applyBlock | rollbackTo` sum, so this property will not hold. Unless we don't generate rollbacks…

Reply:
yes, when rollbacks kick in we will need to compensate this. But, still. those properties should hold. Even, after rollback we should have `slotIdAbsolute` occur only once in values. For each PoolId we should have only non-decreasing number of values (when compared with the state for an earlier header block).

Reply:
nevertheless, I believe having these properties precisely stated would increase confidence the metrics were accounted for rollback properly.